catsoop/lti.py

```diff
@@ -0,0 +1,297 @@
+
+    return login_info
+
+
+def serve_lti(context, path_info, environment, params, dispatch_main):
+    '''
+    context: (dict) catsoop global context
+    path_info: (list) URL path components
+    environment: (dict-like) web server data, such as form input
+    dispatch_main: (proc) call this with environment to dispatch to render URL
+    '''
+    if not 'cs_lti_config' in context:
+        msg = "[lti] LTI not configured - missing cs_lti_config in config.py"
+        LOGGER.error(msg)
+        raise Exception(msg)
+
+    LOGGER.info("[lti] parameters=%s" % params)
+    lti_action = path_info[0]
+    LOGGER.info("[lti] lti_action=%s, path_info=%s" % (lti_action, path_info))
+
+    session = context['cs_session_data']
+    l4c = lti4cs(context, session, {}, {})
+    lti_ok = l4c.verify_request(params, environment)
+    if not lti_ok:
+        msg = "LTI verification failed"
+    else:
+        lti_data = session["lti_data"]
+        uname = "lti_%s" % lti_data.get("lis_person_sourcedid", lti_data['user_id'])
+        email = lti_data.get('lis_person_contact_email_primary', "%s@unknown" % uname)
+        name = lti_data.get('lis_person_name_full', uname)
+        
+        get_or_create_user(context, uname, email, name)
+        l4c.save_lti_data(context)	# save lti data, e.g. for later use by the checker
+        if lti_action=="course":
+
+            LOGGER.info("[lti] rendering course page for %s" % uname)
+
+            sub_path_info = path_info[1:]	# path without _lti/course prefix
+            sub_path = '/'.join(sub_path_info)
+            LOGGER.info("[lti] sub_path=%s" % sub_path)
+            environment['PATH_INFO'] = sub_path
+            environment['session_id'] = context['cs_sid']	# so that a new session ID isn't generated
+            return dispatch_main(environment)
+
+        msg = "Hello LTI"
+
+    return (
+        ("200", "Ok"),
+        {"Content-type": "text/plain", "Content-length": str(len(msg))},
+        msg
+    )
+        
+#-----------------------------------------------------------------------------
+
+def get_new_password_salt(length=128):
+    """
+    Generate a new salt of length length.  Tries to use os.urandom, and
+    falls back on random if that doesn't work for some reason.
+    """
+    try:
+        out = os.urandom(length)
+    except:
+        out = "".join(chr(random.randint(1, 127)) for i in range(length)).encode()
+    return out
+
+def compute_password_hash(context, password, salt=None, iterations=500000, encrypt=True):
+    """
+    Given a password, and (optionally) an associated salt, return a hash value.
+    """
+    hash_ = hashlib.pbkdf2_hmac(
+        "sha512", _ensure_bytes(password), _ensure_bytes(salt), iterations
+    )
+    if encrypt and (context["csm_cslog"].ENCRYPT_KEY is not None):
+        hash_ = context["csm_cslog"].FERNET.encrypt(hash_)
+    return hash_
+
+def _ensure_bytes(x):
+    try:
+        return x.encode()
+    except:
+        return x
```

Comment:
It should be possible to grab `get_new_password_salt`, `compute_password_hash`, and `_ensure_bytes` from the login module directly, with something like the following:

```
_login = auth.get_auth_type_by_name(context, 'login')
```

`_login` will then be a dictionary containing those functions, so it's not necessary to repeat their definitions here.